### PR TITLE
Switch to SBT, added layers

### PR DESCRIPTION
**Pull request by Gaarv**
## What changes were proposed in this pull request?

I planned for quite some time to work on ScalNet for 2018 and while it could be redundant with DL4J, the reduced boilerplate which can be done with Scala could be a great benefit to help beginners. Regardless of level of DL4J knownledge, you can also access everything of DL4J Java API from Scala anyway. Im hoping this could help adoption, even by a little.

Seeing the project hasn't seen many commits those last months, I went ahead and took some freedom, so I would understand if those changes are too much. Nothing was broke API-wise tho or drastically changed.
Before having to resort to make my own fork, I really wanted to submit a PR first to see what you think about it. I know you guys have much more important work to do in DL4j than writing a wrapper for it, so if this could be of help, why not.

The changes :

* Install and build with SBT. I think ScalNet should be more of a Scala project. I tried to work with maven for a couple of weeks on ScalNet but it was really annoying (sbt-pom-reader wasn't working for me BTW). I'm using Maven daily for Java projects but for Scala projects, it just get in the way IMO, just the build configuration is a pain. There's also a lot of nice sbt plugins like "sbt-javacpp" to help with dependencies. However Maven is still required when using snapshots as of now because sbt has still issues with nd4j depencies, like Gradle. I explained that in the README.

* Added layers Embedding, GravesLSTM, RnnOutput. I started from some DL4J examples and made the wrappers from there. It helped me started understanding how things was already implemented in ScalNet.

* Used DL4J/ND4J enums so all optimizers/updaters can be used in layers. On the other hand, they're not configurable yet (ie. learning rate).

* Added model.evaluate() function as helper for evaluator in simple cases.

* Added tests and examples related to added layers. Refactoring, cleaning here and there.

* Autoformat with scalafmt and automatic license headers. When I first cloned the project I saw this was not enforced at compile / test and it could be useful to do so for contributions.

* Changed logging to scala-logging + logback-classic. A light and efficient combo I like to use in my Scala projects, reducing boilerplate when a logger is needed. Configuration provided.

* Support only for Scala 2.11. For a project still in Alpha, 2.10 support shouldn't be mandatory. Packages build for Spark 1.6.3 (already pretty old but latest from 1.X) and Spark 2.X can both use Scala 2.11.

* Since everything has been changed for SBT, if you choose to merge this PR, it will probably need some work on your end to get it back with your release process. I tried to put everything needed in the build.sbt to push into sonatype snapshots / releases using [sbt-git](https://github.com/sbt/sbt-git).

## How was this patch tested?

* Build and tests are ran with [Travis on Github](https://travis-ci.org/Gaarv/ScalNet), removed some examples which were ran in tests because Travis don't have to power to do deep learning, even on MNIST :)

* Ran all examples manually multiple times

## TODO

After this PR I plan to:

* Add more missing layers, more tests. On examples, only adding those that are meaningfull for tests or interesting of usage (model import, Spark, UI). No need to replicate all examples from DL4J.

* A functional API, to be able to use "merge" or build complex graphs.

* Making a choice between Sequential() and NeuralNet() models. I like NeuralNet() more, plus it has inferred preprocessors. Maintaining both could be cumbersome. Opinions ?

* Reducing boilerplate by adding implicits / default listeners like early stopping. They should be configurable.

### CODE REVIEW

**Comment by agibsonccc** [comment]
Oh this is great! Let me put @maxpumperla on this. Thanks again!

**Comment by saudet** [comment]
Please leave files like Jenkinsfile and perform-release.sh!

**Comment by maxpumperla** [comment]
@Gaarv that's great work! I'll get back to you with feedback as soon as I can.

File: .scalafmt.conf [hunk 1]
```diff
@@ -0,0 +1,9 @@
+align = some
+danglingParentheses        = true
```
Comment (maxpumperla) [inline .scalafmt.conf]: 👍

File: Jenkinsfile [hunk 2]
```diff
@@ -1,10 +0,0 @@
-#!groovy
-
-/*
-    To redefine some job/run parameters,
-    please provide arguments to jenkinsBuilder step.
-    Example: jenkinsBuilder platforms: []
```
Comment (maxpumperla) [inline Jenkinsfile]: yeah, as @saudet already mentioned, please put the existing files back in.

File: README.md [hunk 3]
```diff
@@ -1,38 +1,60 @@
...
 
-Currently ScalNet depends on deeplearning4j and nd4j SNAPSHOTs. Please install these package in your local machine first with `mvn install`.
+Currently ScalNet depends on deeplearning4j and nd4j SNAPSHOTS. 
 
 - [deeplearning4j/deeplearning4j](https://github.com/deeplearning4j/deeplearning4j)
 - [deeplearning4j/nd4j](https://github.com/deeplearning4j/nd4j)
 
-Target for scala-2.11
+ScalNet use sbt, but due to [resolving issues](https://nd4j.org/dependencies), you must have maven available to copy some nd4j-native dependencies in classpath, in orderto run the examples.
+
```
Comment (maxpumperla) [inline README.md]: ScalNet use**s**
in **your** classpath
in **order to**

File: buildmultiplescalaversions.sh [hunk 4]
```diff
@@ -1,37 +0,0 @@
-#! /bin/bash
-BASEDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-
-function echoError() {
-    (>&2 echo "$1")
-}
-
```
Comment (maxpumperla) [inline buildmultiplescalaversions.sh]: you're right that 2.10 becomes extinct, but in general (large) companies are slow to adapt. please leave these helper scripts in the project for now.

File: change-scala-versions.sh [hunk 5]
```diff
@@ -1,94 +0,0 @@
-#!/usr/bin/env bash
-
-#
-# Licensed to the Apache Software Foundation (ASF) under one or more
-# contributor license agreements.  See the NOTICE file distributed with
-# this work for additional information regarding copyright ownership.
-# The ASF licenses this file to You under the Apache License, Version 2.0
-# (the "License"); you may not use this file except in compliance with
```
Comment (maxpumperla) [inline change-scala-versions.sh]: this one, too.

File: perform-release.sh [hunk 6]
```diff
@@ -1,56 +0,0 @@
-#!/bin/bash
-set -eu
-
-if [[ $# < 2 ]]; then
-    echo "Usage: bash perform-release.sh release_version snapshot_version staging_repository"
```
Comment (maxpumperla) [inline perform-release.sh]: and especially this one :)

File: pom.xml [hunk 7]
```diff
@@ -1,400 +1,57 @@
...
-      </plugin>
-    </plugins>
-  </reporting>
-
-  <profiles>
-    <profile>
-      <id>release-sign-artifacts</id>
-      <activation>
-        <property>
-          <name>performRelease</name>
```
Comment (maxpumperla) [inline pom.xml]: It's cool that you slimmed this down quite a bit, but it was setup to fit into our CI pipeline.

File: src/main/scala/org/deeplearning4j/scalnet/layers/GravesLSTM.scala [hunk 8]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2016 Skymind
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/layers/GravesLSTM.scala]: I'd love to adopt a folder structure that mimics this one https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/layers

i.e. put this layer under `recurrent`. Also, we have a regular "LSTM" layer by now, which should be no work to integrate as well (and is maybe a bit more relevant).

File: src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/convolution/LeNetMnistExample.scala [hunk 9]
```diff
@@ -1,88 +1,96 @@
...
 
-  log.info("Build model....")
-  private val model: NeuralNet = NeuralNet(
-    inputType = InputType.convolutionalFlat(nbRows, nbColumns, nbChannels), rngSeed = seed)
-  model.add(Convolution2D(nFilter = 20, nChannels = nbChannels, kernelSize = List(5, 5), stride = List(1, 1),
-    weightInit = WeightInit.XAVIER, regularizer = L2(weightDecay)))
+  logger.info("Build model...")
+  val model: NeuralNet = NeuralNet(
+    inputType = InputType.convolutionalFlat(height, width, channels),
+    rngSeed = seed
```
Comment (maxpumperla) [inline src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/convolution/LeNetMnistExample.scala]: Ok, I get that this is just cosmetics, but it's one of the selling points of ScalNet to have a similar fast, very-few-lines prototyping experience like Keras. Putting every arg on a new line makes this _look_ more verbose. I purposefully condensed the examples as much as possible.

Maybe we can find a compromise here?!

File: src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/feedforward/MLPMnistTwoLayerExample.scala [hunk 10]
```diff
@@ -1,83 +1,83 @@
...
-  model.add(Dense(nOut = 500, nIn = numRows * numColumns, weightInit = WeightInit.XAVIER,
-    activation = Activation.RELU, regularizer = L2(learningRate * decay)))
-  model.add(Dense(nOut = 100, weightInit = WeightInit.XAVIER, activation = Activation.RELU,
-    regularizer = L2(learningRate * decay)))
-  model.add(Dense(outputNum, weightInit = WeightInit.XAVIER, activation = Activation.SOFTMAX,
-    regularizer = L2(learningRate * decay)))
+  model.add(
+    Dense(nIn = height * width,
+          nOut = hiddenSize,
+          weightInit = WeightInit.XAVIER,
```
Comment (maxpumperla) [inline src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/feedforward/MLPMnistTwoLayerExample.scala]: same here, I personally favour the concise versions.

File: src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/recurrent/BasicRNNExample.scala [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+import org.nd4j.linalg.factory.Nd4j
+import org.nd4j.linalg.lossfunctions.LossFunctions.LossFunction
+
+import scala.util.Random
+
+object BasicRNNExample extends App with LazyLogging {
+
+  // define a sentence to learn.
+  // Add a special character at the beginning so the RNN learns the complete string and ends with the marker.
+  val learningString = "*Der Cottbuser Postkutscher putzt den Cottbuser Postkutschkasten.".toVector
```
Comment (maxpumperla) [inline src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/recurrent/BasicRNNExample.scala]: Süßes Beispiel! :)

File: src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/recurrent/BasicRNNExample.scala [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    rnn.rnnClearPreviousState()
+    val init = Nd4j.zeros(learningChars.length)
+    init.putScalar(learningChars.indexOf(learningString(0)), 1)
+    var output = rnn.rnnTimeStep(init)
+
+    val predicted: Vector[Char] = learningString.map { _ =>
+      val sampledCharacterIdx = Nd4j.getExecutioner.exec(new IMax(output), 1).getInt(0)
+      val nextInput = Nd4j.zeros(learningChars.length)
+      nextInput.putScalar(sampledCharacterIdx, 1)
+      output = rnn.rnnTimeStep(nextInput)
```
Comment (maxpumperla) [inline src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/recurrent/BasicRNNExample.scala]: That's a great addition, thanks for adding this example.

File: src/test/scala/org/deeplearning4j/scalnet/layers/DenseTest.scala [hunk 13]
```diff
@@ -1,52 +1,67 @@
...
 
 package org.deeplearning4j.scalnet.layers
 
+import org.deeplearning4j.nn.conf.layers.{ DenseLayer, OutputLayer => JOutputLayer }
 import org.nd4j.linalg.lossfunctions.LossFunctions.LossFunction
-import org.deeplearning4j.nn.conf.layers.{DenseLayer, OutputLayer => JOutputLayer}
-import org.scalatest.FunSpec
-
-/**
-  * Created by maxpumperla on 29/06/17.
```
Comment (maxpumperla) [inline src/test/scala/org/deeplearning4j/scalnet/layers/DenseTest.scala]: Go ahead and mark the new files with your name, you're the author after all. (you don't necessarily have to delete mine, though ;))

File: README.md (hunk 3 above)
Comment (Gaarv) [reply README.md]: Oops, sorry about my bad non-native English :) Should have double-checked the readme at least, I will be more careful.

File: buildmultiplescalaversions.sh (hunk 4 above)
Comment (Gaarv) [reply buildmultiplescalaversions.sh]: You're right, it might be too extreme after all. Will be bringing back 2.10 in both maven and sbt.

File: Jenkinsfile (hunk 2 above)
Comment (Gaarv) [reply Jenkinsfile]: It's great to hear your opinion about this. Indeed it's totally feasible to have maven and sb, jenkins and travis. I will add into the readme how to build with both too.

File: src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/convolution/LeNetMnistExample.scala (hunk 9 above)
Comment (Gaarv) [reply src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/convolution/LeNetMnistExample.scala]: In fact, it was more because of autoformatting rules. I wasn't sure how well-received would be examples without args seeing they were everywhere in previous ones so I didn't want to temper it too much. I'm myself a long Keras user and I'm totally for short constructor lines :) Will gladly do it.

File: src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/recurrent/BasicRNNExample.scala (hunk 11 above)
Comment (Gaarv) [reply src/test/scala/org/deeplearning4j/scalnet/examples/dl4j/recurrent/BasicRNNExample.scala]: No idea what it means, just took it from the same DL4J example haha.

File: src/test/scala/org/deeplearning4j/scalnet/layers/DenseTest.scala (hunk 13 above)
Comment (Gaarv) [reply src/test/scala/org/deeplearning4j/scalnet/layers/DenseTest.scala]: Sorry for the offense ^^; I never know if those are put on purpose and sometimes I get into the habit of just removing them. I might have removed a couple before stepping back on this, sorry will try to restore where needed. I use git log to blame or credit people so I usually don't need those xD

Comment (maxpumperla) [reply src/test/scala/org/deeplearning4j/scalnet/layers/DenseTest.scala]: absolutely no offense taken, was more of a joke actually. but if you want, please do add your name. :)

**Comment by Gaarv** [comment]
Thank you for the detailed and constructive review! I now have more precise guidance on how to help the project also for futures PR. I will work on the changes starting tomorrow.

File: src/main/scala/org/deeplearning4j/scalnet/layers/GravesLSTM.scala (hunk 8 above)
Comment (Gaarv) [reply src/main/scala/org/deeplearning4j/scalnet/layers/GravesLSTM.scala]: Sure thing! I first used the same structure than DL4J in order to compare "what's missing" from shell, I even implemented placeholder traits (abstract layers mostly) but eventually decided it was too much dead code.

Actually, I'm already working on those on a local branch ;) (LSTM, Bidirectional)

**Comment by maxpumperla** [comment]
@Gaarv let me know once you're ready, then I'll give it another pass!

**Comment by Gaarv** [comment]
Sure,  almost done, will let you know :) I've been having issues with nd4j-native since a few days and I want to be sure nothing is due to any of my changes. Seems related to https://github.com/deeplearning4j/nd4j/issues/2720

**Comment by Gaarv** [comment]
Okay this should be good. Because of those issues with nd4j-native I added back some integration tests. I ran those extensively and they pass on Windows, Linux Debian https://gist.github.com/Gaarv/731b62972283dbbde100f61f1c757dd0 (fresh Google Cloud instance), but won't on my Fedora and Travis :/

I will probably file an ND4J about this issue but this is unrelated to this PR. Let me know how it goes on your Jenkins.

**Comment by sshepel** [comment]
@Gaarv you can check the result of the build by clicking on the `Details` link in front of `continuous-intergration/jenkins/pr-merge`. 
But it looks like the permissions of the ./change-scala-versions.sh file has been changed and Jenkins failed to run that script.
ATM we do have simple maven command call `mvn -U clean install -Dlocal.software.repository=sonatype -Dmaven.test.skip=true` for the build stage. Does it needs to be changed to be able to build your changes?

**Comment by Gaarv** [comment]
I made the changes requested by @maxpumperla (basically putting back maven build and ensure it run with shell scripts) so you shouldn't have to change anything in your build compared to before :) I just added an extra Travis build on my side so I can run / check tests on sbt and maven alike. 

Sorry for the forgotten chmod +x :) its fixed

**Comment by sshepel** [comment]
> Sorry for the forgotten chmod +x :) its fixed
No worries and thanks a lot.

As for the build changes, that is awesome, I also can see that now the build has been fixed.

Once again, thanks a lot!
